--- Scsi.h	2011-05-27 11:30:15.000000000 +0400
+++ Scsi-new.h	2012-05-15 10:23:15.000000000 +0400
@@ -49,8 +49,10 @@
 #define EFI_SCSI_OP_READ6           0x08
 #define EFI_SCSI_OP_READ10          0x28
 #define EFI_SCSI_OP_READ16          0x88
+#define EFI_SCSI_OP_READ12          0xA8
 #define EFI_SCSI_OP_READ_CAPACITY   0x25
 #define EFI_SCSI_OP_READ_CAPACITY16 0x9e
+#define EFI_SCSI_OP_READ_FORMAT_CAPACITY   0x23
 #define EFI_SCSI_OP_READ_DEFECT     0x37
 #define EFI_SCSI_OP_READ_LONG       0x3e
 #define EFI_SCSI_OP_REASSIGN_BLK    0x07
@@ -64,11 +66,17 @@
 #define EFI_SCSI_OP_SEND_DIAG       0x1d
 #define EFI_SCSI_OP_SET_LIMIT       0x33
 #define EFI_SCSI_OP_START_STOP_UNIT 0x1b
+#define   EFI_SCSI_SUBOP_STOP_DISC    0x0
+#define   EFI_SCSI_SUBOP_START_DISC   0x1
+#define   EFI_SCSI_SUBOP_EJECT_DISC   0x2
+#define   EFI_SCSI_SUBOP_CLOSE_TRAY   0x3
+
 #define EFI_SCSI_OP_SYNC_CACHE      0x35
 #define EFI_SCSI_OP_VERIFY          0x2f
 #define EFI_SCSI_OP_WRITE6          0x0a
 #define EFI_SCSI_OP_WRITE10         0x2a
 #define EFI_SCSI_OP_WRITE16         0x8a
+#define EFI_SCSI_OP_WRITE12         0xaa
 #define EFI_SCSI_OP_WRITE_VERIFY    0x2e
 #define EFI_SCSI_OP_WRITE_LONG      0x3f
 #define EFI_SCSI_OP_WRITE_SAME      0x41
